### PR TITLE
[query] Add PartitionReader for general implementations of ReadPartition.

### PR DESCRIPTION
**Pull request by cseed**
Change was surprisingly easy.  Next:
 - TableReaders should know how to lower themselves to PartitionReaders, making ReadPartition lowering generic, starting with TableNativeReader.
 - Add a PartitionReader that can be implemented by providing a `(context: Any, region: Region) => Iterator[Long]`.
 - Then write PartitionReaders for the various table and matrix readers that can turn into turn into RVDs or be lowered into PartitionReaders.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 1]
```diff
@@ -163,6 +164,45 @@ object TableNativeReader {
   }
 }
 
+case class PartitionNativeReader(spec: AbstractTypedCodecSpec, fullRowType: TStruct) extends PartitionReader {
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala:167]: Small note: is there a reason the native partition reader explicitly takes the full row type in the constructor? It should be defined as `encodedVirtualType` in the TypedCodecSpec.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala:167]: this is the requested type probably

Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala:167]: As I read it, `fullRowType` is the type of the encoded data, and the requested type is passed in as arguments to various methods when needed and isn't stored with the reader.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala:167]: oh, I wasn't reading closely enough. I think you're right.

Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: Arcturus was right, I can get the full type from the spec.  Fixed. Thanks!

**Pull request merged**